### PR TITLE
add blog "how to keep safe as a client"

### PR DESCRIPTION
**Pull request by LeasyZhang**
writing a blog about Circuit Breaker

### CODE REVIEW

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+date:   2019-01-27 22:00:00 +0800
+categories: blog
+author: Joe Zhang
+---
+
+Recently I am working on a project called transcript, It will use many third party services, generally third party services will expose restful APIs and my project will act as a client.
+
+![](https://github.com/BarkZhang/integration-blog/blob/master/assets/2019-01-24-how-to-keep-safe-as-a-client/transcript_sample.png)
+
+In such background, If some services broken down or having high latency, It will affect our system. what's more, It will block user request and may cause the whole system broken down.
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: If some services broken down -> If some services were broken

Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: If some services were broken-down / if some services were broken

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+In order to resolve such problem, one solution is using Circuit Breaker.
+
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vSd4AzVWGyFNonEC5CuA1A09St2nUJ20ViaT2CPWl9GEBzuiazRoasZiaKQ/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: I think it would be better to add the reference if we adopt diagrams from others directly

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+## What is Circuit Breaker
+
+Circuit Breaker is a protection mechanism. Like switch blade in electronic system, If there are something wrong with one part, Circuit breaker will open and cut off electric. A Circuit Breaker in a system means: if downstream service is broken down for some reason, upsrteam system will cut off connections with downsream services.
+
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: typo 'downsream' -> downstream

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md (hunk 1 above)
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: We should use initial capital right after period(.) and lower case after comma(,)
In such background, if some services broken down or having high latency, it will affect our system. What's more, it will block user request and may cause the whole system broken down.

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+if(isOpenCircuitBreaker == true){
+  return fail;
+}
+
+// we invoke remote service here
+
+```
+### Try to reconnect
+
+After cut off connections, We need to recover as soon as possbile, Because currently the system is not complete. In order to automatically recover connection we need to define a standard, like previously how we judge a system is unavailable, we need to define whether a system is abaliable again.
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: abaliable -> available

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+- Threshold, for example, there are 100 success request within 5 seconds.
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: request -> requests

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+- Percentage, for example, there are 95% success request within 10 seconds.
+
+It also contains time window, threshold and percentage. But there is a slight difference: most of the time, the system will keep unavailable for a momment, So we can check in a fixed times(30s for example).
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: momment -> moment

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  errorCount++;
+  if(errorCount == threshold){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+However reconnect may failed, In a large system we can't use all servers to try reconnecting. We can take these measures:
+- 1.Use a few servers to check
+- 2.Add a health check interface to judge system conditions: CPU, IO for example.
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: judge -> check

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+### recover
+
+Once the system is avaliable, we need to recover.
+
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vSUGPwicRFU6xJeNPxzKOWEhlM3yKSVdEfk1nusm5FfnRm1WyYkOVybBg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+
+This is the general process of Circuit Breaker, Frameworks like Hystrix, resilience4j are all have such process flow.
```
Comment (alvita) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: ... are all have such process flow -> ... are designed based on such process flow (concept)

Comment (BarkZhang) [reply _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: Thanks alvita! you are so patient and careful! I updated these spelling mistakes and add a reference article!

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
+---
+layout: post
+title:  "how to keep safe as a client"
+date:   2019-01-27 22:00:00 +0800
+categories: blog
+author: Joe Zhang
+---
+
+Recently I am working on a project called transcript, It will use many third party services, generally third party services will expose restful APIs and my project will act as a client.
```
[the remark this answers is not in the document]
Comment (BarkZhang) [reply _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: You are right, I update my blog and use a virtual project instead of the real name, and remove the image about transcript because actually I don't need to.
Thanks!!

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+layout: post
+title:  "how to keep safe as a client"
+date:   2019-01-27 22:00:00 +0800
+categories: blog
+author: Joe Zhang
+---
+
+[Artical Link](https://mp.weixin.qq.com/s?__biz=MzAxODcyNjEzNQ==&mid=2247486481&idx=1&sn=87aee20e301d87030be2636cd0a124b7&chksm=9bd0a189aca7289f0a5e8a91907d21e32bd367341251c713e76c2fd97f6f64c06379ad7c4f93&scene=21#wechat_redirect): 
+
+Previously I worked on a project which use many third party services, These third party services exposed restful APIs and my project will act as a client.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: Previously I worked on a project *that used* many..., *these* third party... and my project *acted* as a client.

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+In such background, If some services were broken or having high latency, It will affect our system. what's more, It will block user request and may cause the whole system broken down.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: Ok, the general rule is that a period (.) denotes the end of sentence, so you capitalize the next word that starts the next sentence. But a comma (,), split a sentence into smaller parts, you don't need to capitalize the word after commas.

Comment (LeasyZhang) [reply _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: thanks jack, I just fixed grammar issues and capitalize issues.

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vSUGPwicRFU6xJeNPxzKOWEhlM3yKSVdEfk1nusm5FfnRm1WyYkOVybBg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+
+This is the general process of Circuit Breaker, frameworks like hystrix, resilience4j are all have such process flow.
+
+## Best practice for Circuit Breaker
+
+### What kind of situation require Circuit Breaker
+
+Circuit Breaker is not suitable in every situations. There are some conditions we can consider using Circuit Breaker:
+- The system we rely on is a shared system, and our client is just one of them. Other systems may abuse the shared system and affect us.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: Let me rephrase this a bit:
- If the target system is a shared service: when other systems abuse this shared service, our client can be affected.

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+- Our system is not used in a regular base, somethings the system may used very frequently while other times it is rarely used, we need to hold on the sudden flow.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: This sounds rather confusing, let me try to rephrase it a bit:
- The target system is not in regular use, sometimes the system may be under heavy load, while other times it may be rarely invoked, but we need to maintain stable service during sudden spikes in load.

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+### Pay attention to implement a Circuit Breaker
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: How about:
Things to watch out for when implementing a circuit breaker

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+Circuit Breaker is not perfect, we need to pay attention to two things.
+- Firstly, if we rely on a system that has many replicas, we need to keep in mind that exception in partial node does not mean exception of the whole system.
+- Secondly, Circuit Breaker should be the last choice, we can to use "fall back" or "rate limit" prior than Circuit Breaker. Because partial is better than none, although some users may find out service broke, we can try to keep our system working and serve other users. For example, abandon non-core services, or give a friendly hint.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: last choice => last resort
prior than => prior to
find out service broke => find out that the service is broken
For example, abandon ... ... => For example, ```by abandoning``` non-core services, or ``by giving friendly hints```, we can keep our core services working.

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+## Conclusion
+
+Circuit Breaker can be implemented before actual function or after actual function. It is very correspond to AOP concept, so generally most circuit breaker framework use AOP as an implement strategy.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: Please give the full phrase when you use acronyms for the first time to avoid confusion.
For example, the first time I use OOP (object-oriented programming) in my article, I'd put the full phrase in parenthesis.
circuit breaker framework => circuit breaker ```frameworks```

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+Circuit Breaker is a shell, which protect itself when error occurs. But in the long term is better to have regular pressure testing and understand limitations of our system, firstly we try "fall back" or "rate limit" then we can use Circuit Breaker.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: Let's rephrase this a bit:
Circuit breaker is a shell ```that protects the system``` when error occurs. But in the long ```run, it``` is better to have regular pressure tests and ```better understanding of the``` limitations of our system. ```We should try "fall backs" or "rate limits" before we implement circuit breakers.```

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md (hunk 14 above)
Comment (LeasyZhang) [reply _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: Is "watch out for" a fixed collocation or we can remove "for"?

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md (hunk 17 above)
Comment (LeasyZhang) [reply _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: The whole paragraph seems much smoother after rephrase! Thanks again for such detailed review.

**Comment by LeasyZhang** [comment]
@u9520107 Thanks for your detailed review, I fixed these mistakes and checked again using grammarly:)

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+  return success;
+}else{
+  errorCount ++;
+  if(errorCount == threshold){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+Since the reconnection can still fail, we cannot use all servers to try to reconnect in a large system.. We can take these measures:
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: you have an extra '.': in a large system```..```

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+- 1. Use a few servers to check
+- 2. Add a health check interface to check system conditions: CPU, IO etc.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: CPU, IO etc. => CPU, IO```,``` etc.

File: _posts/2019-01-24-how-to-keep-safe-as-a-client.md [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+
+### Things to watch out for when implementing a circuit breaker
+
+Circuit Breaker is not perfect, we need to pay attention to two things.
+- Firstly, if we rely on a system that has many replicas, we need to keep in mind that the exception in a partial node does not mean the exception of the whole system.
+- Secondly, Circuit Breaker should be the last resort, we can to use "fall back" or "rate limit" prior to Circuit Breaker. Because partial is better than none, although some users may find out that the service is broken, we can try to keep our system working and serve other users. For example, by abandoning non-core services, or by giving friendly hints, we can keep our core services working.
+
+## Conclusion
+
+Circuit Breaker can be implemented before the actual function or after the actual function. It meets the AOP(aspect-oriented programming) concept, so generally, most circuit breaker frameworks use AOP(aspect-oriented programming) as an implemented strategy.
```
Comment (u9520107) [inline _posts/2019-01-24-how-to-keep-safe-as-a-client.md]: You can just add "(aspect-oriented programming)" after the first use of AOP. For later uses people will know what it means already. So I think it's better to remove the second "(aspect-oriented programming).
